modules/mbedtls_curl/http_client_curl.cpp

```diff
@@ -0,0 +1,440 @@
+	RequestContext *ctx = (RequestContext *)userdata;
+	String s((const char *)buffer);
+	Vector<String> parts = s.split(":");
+	if (parts.size() != 2) {
+		return size * nitems;
+	}
+
+	ctx->response_headers->push_back(s);
+
+	String header = parts[0].to_lower();
+	String val = parts[1];
+	// Trim any whitespace prefix
+	while (val.begins_with(" ")) {
+		val = val.trim_prefix(" ");
+	}
+
+	// Use the content length to determine body size.
+	if (header == "content-length") {
+		*(ctx->body_size) = val.to_int();
+	}
+
+	// If the Connection header is set to "close" then
+	// keep-alive isn't enabled.
+	if (header == "connection" && val == "close") {
+		*(ctx->keep_alive) = false;
+	}
+
+	if (header == "transfer-encoding" && val == "chunked") {
+		*(ctx->body_size) = -1;
+		*(ctx->chunked) = true;
+	}
+
+	*(ctx->has_response) = true;
+
+	return size * nitems;
+}
+
+size_t HTTPClientCurl::_read_callback(char *buffer, size_t size, size_t nitems, void *userdata) {
+	RingBuffer<uint8_t> *b = (RingBuffer<uint8_t> *)userdata;
+	int n = b->copy((uint8_t *)buffer, 0, size * nitems);
+	return n;
+}
+
+size_t HTTPClientCurl::_write_callback(char *buffer, size_t size, size_t nitems, void *userdata) {
+	RequestContext *ctx = (RequestContext *)userdata;
+	PackedByteArray chunk;
+	chunk.resize(size * nitems);
+	memcpy(chunk.ptrw(), buffer, size * nitems);
+	ctx->response_chunks->append(chunk);
+	*(ctx->status) = STATUS_BODY;
+
+	return size * nitems;
+}
+
+curl_slist *HTTPClientCurl::_ip_addr_to_slist(const IPAddress &p_addr) {
+	String addr = String(p_addr);
+	if (addr.find(":") != -1) {
+		addr = "[" + addr + "]";
+	}
+	String h = host + ":" + String::num_int64(port) + ":" + addr;
+	return curl_slist_append(nullptr, h.ascii().get_data());
+}
+
+String HTTPClientCurl::_hostname_from_url(const String &p_url) {
+	String hostname = p_url.trim_prefix("http://");
+	hostname = hostname.trim_prefix("https://");
+	return hostname.split("/")[0];
+}
+
+Error HTTPClientCurl::_resolve_dns() {
+	IP::ResolverStatus rstatus = IP::get_singleton()->get_resolve_item_status(resolver_id);
+	switch (rstatus) {
+		case IP::RESOLVER_STATUS_WAITING:
+			return OK;
+		case IP::RESOLVER_STATUS_DONE: {
+			IPAddress addr = IP::get_singleton()->get_resolve_item_address(resolver_id);
+
+			Error err = _request(addr, true);
+
+			IP::get_singleton()->erase_resolve_item(resolver_id);
+			resolver_id = IP::RESOLVER_INVALID_ID;
+
+			if (err != OK) {
+				status = STATUS_CANT_CONNECT;
+				return err;
+			}
+			return OK;
+		} break;
+		case IP::RESOLVER_STATUS_NONE:
+		case IP::RESOLVER_STATUS_ERROR: {
+			IP::get_singleton()->erase_resolve_item(resolver_id);
+			resolver_id = IP::RESOLVER_INVALID_ID;
+			close();
+			status = STATUS_CANT_RESOLVE;
+			return ERR_CANT_RESOLVE;
+		} break;
+	}
+
+	return OK;
+}
+
+Error HTTPClientCurl::_poll_curl() {
+	CURLMcode rc = curl_multi_perform(curl, &still_running);
+	if (still_running) {
+		rc = curl_multi_wait(curl, nullptr, 0, 1000, nullptr);
```

Comment:
This is blocking (`1000` is the timeout). Should probably be:
```
curl_multi_wait(curl, nullptr, 0, 0, nullptr)
```

And I think we can use this to implement the blocking mode.

Reply:
@Faless okay this is done. It seems to work as expected.